File: 动态规划/64.最小路径和.py
class Solution:
    # 一行一行遍历求每个位置
    def minPathSum(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        # 计算边界上的每个点到起点的开销，之后每个空格都可以计算出来
        for i in range(1, m):
            grid[i][0] += grid[i - 1][0]
        for i in range(1, n):
            grid[0][i] += grid[0][i - 1]
        
        # 根据 左边和上边计算当前位置的最小花费
        for i in range(1, m):
            for j in range(1, n):
                grid[i][j] += min(grid[i - 1][j], grid[i][j - 1])

        return grid[-1][-1]

File: 动态规划/64.最小路径和.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function (grid) {
  const rows = grid.length, cols = grid[0].length;
  //初始化
  for (let i = 1; i < rows; i++) grid[i][0] += grid[i - 1][0];
  for (let j = 1; j < cols; j++) grid[0][j] += grid[0][j - 1];

  for (let i = 1; i < rows; i++) {
    for (let j = 1; j < cols; j++) {
      grid[i][j] = Math.min(grid[i - 1][j], grid[i][j - 1]) + grid[i][j];
    }
  }
  return grid[rows - 1][cols - 1];
};
